File: javascript/lasagna-master/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

/**
 * Implement the functions needed to solve the exercise here.
 * Do not forget to export them so they are available for the
 * tests. Here an example of the syntax as reminder:
 *
 * export function yourFunction(...) {
 *   ...
 * }
 */

export function cookingStatus(remainingTime) {
    switch (remainingTime) {
        case undefined:
            return 'You forgot to set the timer.'
        case 0:
            return 'Lasagna is done.'
        default:
            return 'Not done, please wait.';
    }
}

export function preparationTime(layers, timePerLayerInMinutes = 2) {
    return layers.length * timePerLayerInMinutes;
}

export function quantities(layers) {
    return layers.reduce((acc, layer) => {
        if (layer === 'noodles') {
            acc.noodles += 50;
        }
        if (layer === 'sauce') {
            acc.sauce += 0.2;
        }
        return acc;
    }, { noodles: 0, sauce: 0 });
}

export function addSecretIngredient(friendsList, myList) {
    myList.push(friendsList[friendsList.length - 1]);
}

export function scaleRecipe(recipe, portions) {
    return Object.fromEntries(
        Object.entries(recipe).map(([k, v]) => [k, v * portions / 2])
    );
}

File: go/lasagna-master/lasagna_master.go
package lasagna

// TODO: define the 'PreparationTime()' function
func PreparationTime(layers []string, preTimePerLayer int) int {
	if preTimePerLayer == 0 {
		preTimePerLayer = 2
	}
	return len(layers) * preTimePerLayer
}

// TODO: define the 'Quantities()' function
func Quantities(layers []string) (int, float64) {
	gramOfNoodles, literOfSauce := 0, 0.0
	for _, layer := range layers {
		if layer == "noodles" {
			gramOfNoodles += 50
		}
		if layer == "sauce" {
			literOfSauce += 0.2
		}
	}
	return gramOfNoodles, literOfSauce
}

// TODO: define the 'AddSecretIngredient()' function
func AddSecretIngredient(friendsList []string, myList []string) {
	myList[len(myList)-1] = friendsList[len(friendsList)-1]
}

// TODO: define the 'ScaleRecipe()' function
func ScaleRecipe(quantities []float64, portions int) []float64 {
	newQuantites := make([]float64, len(quantities))
	for i, quantity := range quantities {
		newQuantites[i] = quantity * float64(portions) / 2
	}
	return newQuantites
}

// Your first steps could be to read through the tasks, and create
// these functions with their correct parameter lists and return types.
// The function body only needs to contain `panic("")`.
//
// This will make the tests compile, but they will fail.
// You can then implement the function logic one by one and see
// an increasing number of tests passing as you implement more
// functionality.
